File: test/AllPossibleTrajectories.py
import turtle
import reeds_shepp as rs
import utils
import draw
import math
import random as rd
import sys
import os



def main():
    firstAngle = 0
    lastAngle = 0
    pts = []

    with open("route.txt", "r") as reader:
        # Read and print the entire file line by line
        line = reader.readline()
        firstAngle = float(line)
        line = reader.readline()
        lastAngle = float(line)
        i = 0
        prev = 0
        line = reader.readline()
        while line != '':  # The EOF char is an empty string
            num = float(line)
            if i % 2 == 1:
                pts.append((prev/ rs.PathElement.RATIO, num/ rs.PathElement.RATIO))
            line = reader.readline()
            i = i + 1
            prev = num

    
    # generate PATH so the vectors are pointing at each other
    PATH = []
    PATH.append((pts[0][0], pts[0][1], firstAngle))
    for i in range(1, len(pts) - 1):
        dx = pts[i+1][0] - pts[i][0]
        dy = pts[i+1][1] - pts[i][1]
        theta = math.atan2(dy, dx)
        PATH.append((pts[i][0], pts[i][1], utils.rad2deg(theta)))
    PATH.append((pts[-1][0], pts[-1][1], lastAngle))

    
    # init turtle
    tesla = turtle.Turtle()
    tesla.speed(0) # 0: fast; 1: slow, 8.4: cool
    tesla.shape('arrow')
    tesla.resizemode('user')
    tesla.shapesize(1, 1)

    screen = turtle.Screen()
    screen.screensize(2500, 4000)


    # draw vectors representing points in PATH
    for pt in PATH:
        draw.goto(tesla, pt)
        draw.vec(tesla)

    # draw all routes found
    tesla.speed(0)
    #edges = rs.get_all_polygons()

    f = open("trajectory.txt","w")
    
    s = "#Paths " + str(len(PATH) - 1) + "\n"
    f.write(s)#number of segments at all
    #allBannedTrajectories = []
    for i in range(len(PATH) - 1):
        paths = rs.get_all_paths(PATH[i], PATH[i+1])
        #forbiddenTrajectories = []
        s = "_PathID " + str(i) + " from " + str(PATH[i][0]) + " " + str(PATH[i][1]) + " to " + str(PATH[i+1][0]) + " " + str(PATH[i+1][1]) + "\n"
        f.write(s)
        s = "__#Possible_Path_Segment " + str(len(paths)) + "\n"
        f.write(s)
        j = 0
        for path in paths:
            s = "___SegmentID_" + str(j) + "_PathID_" + str(i) 
            
            draw.set_random_pencolor(tesla)
            draw.goto(tesla, PATH[i])
            #first = PATH[i]
            #s = "____ " + str(PATH[i][0]) + " " + str(PATH[i][1]) + " " + str(PATH[i][2]) + "\n"
            
            localTrajectory = draw.draw_path(tesla, path)
            s = s + " " + str(len(localTrajectory)) + "\n"
            f.write(s)
            j += 1

            for k in range(len(localTrajectory)):
                s = "_____Section " + str(round(localTrajectory[k][0], 3)) + " "
                s = s + str(round(localTrajectory[k][1], 3)) + " "
                param = round(localTrajectory[k][2], 2)
                s = s + str(param) + " "
                typeOfTraj = localTrajectory[k][3]
                s = s + str(typeOfTraj) + " "
                steering = localTrajectory[k][4]
                s = s + str(steering) + "\n"
                f.write(s)
    
    f.close()
    # draw shortest route
    tesla.pencolor(1, 0, 0)
    tesla.pensize(3)
    tesla.speed(10)
    draw.goto(tesla, PATH[0])
    path_length = 0
    trajectory = []
    g = open("Trace.txt","w")
    s = str(round(rs.PathElement.RATIO*pts[0][0], 3)) + " " + str(round(rs.PathElement.RATIO*pts[0][1], 3)) + " B\n"
    g.write(s)

    for i in range(len(PATH) - 1):
        path = rs.get_optimal_path(PATH[i], PATH[i+1])
        path_length += rs.path_length(path)
        trajectory = draw.draw_path(tesla, path)
        for j in range(len(trajectory)):
            s = str(round(rs.PathElement.RATIO*trajectory[j][0], 3)) + " " + str(round(rs.PathElement.RATIO*trajectory[j][1], 3)) + \
                " " + str(round(trajectory[j][2], 2)) + " " + trajectory[j][3] + " " + trajectory[j][4] + "\n"
            g.write(s)


    print("Shortest path length: {} px.".format(int(draw.scale(path_length))))
    g.close()

    turtle.done()


if __name__ == '__main__':
        main()

File: test/AllPossibleTrajectories.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <cmath>
#include <vector>
#include <functional>
#include <algorithm>
#include <unistd.h>
#include <fstream>

#include "drawTurtle.cpp"

using namespace std;

void runAllPossibleTrajectories()
{
  
  float firstAngle = 0;
  float lastAngle = 0;

  vector<pair<double, double>> pts;
  cout << "============================== SET UP =========================" << endl;
  fstream f;
  f.open("route.txt", ios::in);
  if (f.is_open())
  {
    string line;
    getline(f, line);
    firstAngle = stod(line);
    getline(f, line);
    lastAngle = stod(line);

    int i = 0;
    double prev = 0;
    getline(f, line);
    while (!f.eof())
    {
      float num = stod(line);
      if (i % 2 == 1)
      {
        pts.push_back(make_pair(prev / PathElement::RATIO, num / PathElement::RATIO));
      }
      getline(f, line);
      i += 1;
      prev = num;
    }
    f.close();
  }

  vector<Point3> PATH;
  PATH.push_back(Point3(pts[0].first, pts[0].second, firstAngle));

  for (int i = 1; i < pts.size() - 1; i++)
  {
    double dx = pts[i + 1].first - pts[i].first;
    double dy = pts[i + 1].second - pts[i].second;
    double theta = atan2(dy, dx);
    PATH.push_back(Point3(pts[i].first, pts[i].second, rad2deg(theta)));
  }
  PATH.push_back(Point3(pts[pts.size() - 1].first, pts[pts.size() - 1].second, lastAngle));

  // cout << "============================== COMPUTE PATH:  " << PATH.size() << "=========================" << endl;

  fstream trajectoriesFile;
  trajectoriesFile.open("trajectory.txt", ios::out);
  if (trajectoriesFile.is_open())
  {
    cout << "+++ Open create trajectory.txt" << endl;
    vector<TRAJECTORY> localTrajectory;
    string s = "#Paths " + to_string(PATH.size() - 1) + "\n";
    trajectoriesFile << s;
    for (int i = 0; i < PATH.size() - 1; i++)
    { 
      cout << "PATH thứ : " << i << endl;
      vector<vector<PathElement>> paths = get_all_paths(PATH[i], PATH[i + 1]);
      s = "_PathID " + to_string(i) + " from " + to_string(PATH[i].x) + " " + to_string(PATH[i].y) + " to " + to_string(PATH[i + 1].x) + " " + to_string(PATH[i + 1].y) + "\n";
      trajectoriesFile << s;
      s = "__#Possible_Path_Segment " + to_string(paths.size()) + "\n";
      trajectoriesFile << s;
      for (int j = 0; j < paths.size(); j++)
      {
        s = "___SegmentID_" + to_string(j) + "_PathID_" + to_string(i);

        localTrajectory = draw_path(paths[j], PATH[i]);
        s = s + " " + to_string(localTrajectory.size()) + "\n";
        trajectoriesFile << s;

        for (int k = 0; k < localTrajectory.size(); k++)
        {
          s = "_____Section " + to_string(localTrajectory[k].x) + " ";
          s = s + to_string(localTrajectory[k].y) + " ";
          double param = round(localTrajectory[k].param * 100) / 100;
          s = s + to_string(param) + " ";
          string typeOfTraj = localTrajectory[k].typeOfTraj;
          s = s + typeOfTraj + " ";
          string steering = localTrajectory[k].typeOfSteering;
          s = s + steering + "\n";
          trajectoriesFile << s;
        }
      }
    }
    trajectoriesFile.close();
  }
  // cout << "============================== WRITE TRAJECTORIES PATH DONE =========================" << endl;
  fstream traceFile;
  traceFile.open("Trace.txt", ios::out);
  if (traceFile.is_open())
  {
    cout << "+++ Open create trajectory.txt" << endl;
    vector<PathElement> full_path;
    double total_length = 0;
    vector<TRAJECTORY> trajectory;
    string s = to_string(round(PathElement::RATIO * pts[0].first * 1000) / 1000) + " " + to_string(round(PathElement::RATIO * pts[0].second * 1000) / 1000) + " B\n";
    ;
    traceFile << s;
    for (int i = 0; i < PATH.size() - 1; i++)
    {
      vector<PathElement> path = get_optimal_path(PATH[i], PATH[i + 1]);

      for (int j = 0; j < path.size(); j++)
      {
        full_path.push_back(path[j]);
      }

      trajectory = draw_path(path, PATH[i]);
      for (int i = 0; i < trajectory.size(); i++)
      {
        s = to_string(PathElement::RATIO * trajectory[i].x) + " " + to_string(PathElement::RATIO * trajectory[i].y) + " " + to_string(trajectory[i].param) + " " + trajectory[i].typeOfTraj + " " + trajectory[i].typeOfSteering + "\n";
        traceFile << s;
      }

      total_length += path_length(path);
    }

    cout << "Shortest path length: " << total_length << endl;

    // for (int i = 0; i < full_path.size(); i++)
    // {
    //   cout << full_path[i].repr() << endl;
    // }
    traceFile.close();
  }
  cout << "============================== WRITE TRACE , END RUN =========================" << endl;
}

// int main(void)
// {
//   runAllPossibleTrajectories();
// }
